src/rules/no_constant_condition.rs

```diff
@@ -0,0 +1,443 @@
+    assert_lint_ok::<NoConstantCondition>(r#"if ((key || 'k') in obj) {}"#);
+    assert_lint_ok::<NoConstantCondition>(
+      r#"if ((foo || {}) instanceof obj) {}"#,
+    );
+
+    assert_lint_ok::<NoConstantCondition>(r#"if ('' + [y] === '' + [ty]) {}"#);
+    assert_lint_ok::<NoConstantCondition>(r#"if ('a' === '' + [ty]) {}"#);
+    assert_lint_ok::<NoConstantCondition>(r#"if ('' + [y, m, d] === 'a') {}"#);
+    assert_lint_ok::<NoConstantCondition>(
+      r#"if ('' + [y, 'm'] === '' + [ty, 'tm']) {}"#,
+    );
+    assert_lint_ok::<NoConstantCondition>(
+      r#"if ('' + [y, 'm'] === '' + ['ty']) {}"#,
+    );
+    assert_lint_ok::<NoConstantCondition>(
+      r#"if ([,] in
+
+        ($2))
+         ;
+         else
+          ;"#,
+    );
+    assert_lint_ok::<NoConstantCondition>(r#"if ([...x]+'' === 'y'){}"#);
+
+    // TODO(humancalico) add a configuration option for { checkLoops: false } https://eslint.org/docs/rules/no-constant-condition#checkloops
+    // assert_lint_ok::<NoConstantCondition>(r#"while(true);"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"for(;true;);"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"do{}while(true)"#);
+
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo(){while(true){yield 'foo';}}"#,);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo(){for(;true;){yield 'foo';}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo(){do{yield 'foo';}while(true)}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo(){while (true) { while(true) {yield;}}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() {for (; yield; ) {}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() {for (; ; yield) {}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() {while (true) {function* foo() {yield;}yield;}}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() { for (let x = yield; x < 10; x++) {yield;}yield;}"#);
+    // assert_lint_ok::<NoConstantCondition>(r#"function* foo() { for (let x = yield; ; x++) { yield; }}"#);
+
+    assert_lint_err::<NoConstantCondition>(r#"for(;true;);"#, 5);
+    assert_lint_err::<NoConstantCondition>(r#"for(;``;);"#, 5);
+    assert_lint_err::<NoConstantCondition>(r#"for(;`foo`;);"#, 5);
+    assert_lint_err::<NoConstantCondition>(r#"for(;`foo${bar}`;);"#, 5);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(true)"#, 10);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(t = -2)"#, 10);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(``)"#, 10);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(`foo`)"#, 10);
+    assert_lint_err::<NoConstantCondition>(r#"do{}while(`foo${bar}`)"#, 10);
+
+    assert_lint_err::<NoConstantCondition>(r#"true ? 1 : 2;"#, 0);
+    // FIXME(humancalico) Is it supposed to be on column 4
+    assert_lint_err::<NoConstantCondition>(r#"q = 0 ? 1 : 2;"#, 4);
```

Comment:
Which column should it be on?

Reply:
Just checked it on https://eslint.org/demo and it to be correct. I'll remove the FIXME tag.